### DISCUSSION TITLE
Is it possible to create SPARQL queries programmatically using Spargebra?

### DISCUSSION THREAD

**Discussion by labra** [oxigraph/oxigraph#901] [Q&A]
I wanted to create SPARQL queries programmatically in Rust. I was initially planning to use Spargebra but looking to the documentation, I think the only way to create SPARQL queries is parsing them. Is there another way? 

My idea is to do something similar to what can be done in Apache Jena as in [this example](https://github.com/apache/jena/blob/main/jena-examples/src/main/java/arq/examples/algebra/AlgebraExec.java)

**Comment by Tpt** [comment]
Hey! You can build the query programmatically by just initializing the structs and enums by hand starting with [`Query`](https://docs.rs/spargebra/latest/spargebra/enum.Query.html). It's a bit cumbersome but it works. If you feel the need it would be great to add helper functions.

**Comment by hobofan** [comment]
So there is no builder pattern or something similar as you'll find in many other SPARQL libraries.

You can construct queries by just using plain Rust constructors. The main struct that is relevant here is [GraphPattern](https://docs.rs/spargebra/latest/spargebra/algebra/enum.GraphPattern.html), which you'll ultimately wrap in a [Query](https://docs.rs/oxigraph/0.3.22/oxigraph/sparql/struct.Query.html).

In order to execute the query against a oxigraph store, you'll have the convert the `spargebra::Query` into a `oxigraph::Query` via `.into()`.

**Comment by labra** [comment]
Thanks!, and is it possible to declare the prefix declarations? 

I noticed that after parsing a query with prefix declarations and displaying it, the prefixed nodes are converted to full IRIs, and it seems that the prefix declarations are not part of the query anymore, right? 

For my use case, I would like to create a _nice_ SPARQL query with prefix declarations, but I think I can't do it programmatically with Spargebra, is there some way that I can't see?

**Reply by hobofan** [reply]
Yeah, I don't think spargebra is currently well suited for that. You'll also get a rather "verbose" query with e.g. all bound variables spelled out in places where `*` would produce a similar result, as the AST doesn't track ambigous rendering variants.

**Reply by Tpt** [reply]
Yes, spargebra internal representation is done to match the [SPARQL algebra](https://www.w3.org/TR/sparql11-query/#sparqlQuery) (hence its name). Hence it applies the normalizations suggested by the SPARQL spec. I wrote this crate for Oxigraph SPARQL implementation.

I believe nice SPARQL building should be the focus on an other library, trying to do both with spargebra is going to be painful.
